core/controllers/question_test.py

```diff
@@ -111,3 +111,216 @@ def test_get(self):
+
+        collection_services.record_played_exploration_in_collection_context(
+            self.new_user_id, self.collection_id, self.exp_id)
+        change_list = [{'cmd': 'update_question_property',
+                        'property_name': 'title',
+                        'new_value': 'ABC',
+                        'old_value': 'A Question'}]
+        self.payload['change_list'] = json.dumps(change_list)
+        self.payload['commit_message'] = 'update title'
+        self.payload['collection_id'] = self.collection_id
+        self.payload['question_id'] = question_id
+        self.login(self.NEW_USER_EMAIL)
+        response_json = self.put_json(
+            '%s' % feconf.QUESTION_DATA_URL, self.payload,
+            expect_errors=False)
+        self.assertIn('question_id', response_json.keys())
+
+        del self.payload['change_list']
+        response = self.testapp.put(
+            '%s' % feconf.QUESTION_DATA_URL, {'payload': (
+                json.dumps(self.payload))}, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+        del self.payload['commit_message']
+        response = self.testapp.put(
+            '%s' % feconf.QUESTION_DATA_URL, {'payload': (
+                json.dumps(self.payload))}, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+        del self.payload['collection_id']
+        response = self.testapp.put(
+            '%s' % feconf.QUESTION_DATA_URL, {'payload': (
+                json.dumps(self.payload))}, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+        del self.payload['question_id']
+        response = self.testapp.put(
+            '%s' % feconf.QUESTION_DATA_URL, {'payload': (
+                json.dumps(self.payload))}, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+    def test_delete(self):
+        question = question_domain.Question(
+            'dummy', 'A Question',
+            exp_domain.State.create_default_state('ABC').to_dict(), 1,
+            self.collection_id, 'en')
+
+        question_id = question_services.add_question(
+            self.new_user_id, question)
+        question_services.add_question_id_to_skill(
+            question_id, self.collection_id, self.skill_id,
+            self.new_user_id)
+        self.login(self.NEW_USER_EMAIL)
+        response = self.testapp.delete(
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id, question_id),
+            expect_errors=False)
+        self.assertEqual(response.status_int, 200)
+
+        response = self.testapp.delete(
+            '%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id),
+            expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+        response = self.testapp.delete(
+            '%s/' % feconf.QUESTION_DATA_URL, expect_errors=True)
+        self.assertEqual(response.status_int, 405)
```

Comment:
I am not sure how to test the case where collection id and question id are not provided in this setting. I tried %s -> feconf.QUESTION_DATA_URL, which throws a 500 for some reason. :/

Reply:
500 usually means "look at the stack trace of the error". I think you'll need to find a way to view the stack trace, since requests to handlers should never return 500 in general.

The simplest way I can think of to do it would be to add a call from the frontend of a random page (say, the About page or the library) that calls the given endpoint with an $http['delete'](...), then start up a dev server and look at the server logs to see what the 500 error looks like.

Reply:
Thanks @seanlip . Fixed it.